// piral-core/lib/state/createActions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createActions = exports.includeActions = void 0;
const tslib_1 = require("tslib");
const actions = tslib_1.__importStar(require("../actions"));
const app_codegen_1 = require("../../app.codegen");
function createContext(state, events) {
    const ctx = {
        ...events,
        apis: {},
        converters: {
            html: ({ component }) => component,
        },
        navigation: (0, app_codegen_1.createNavigation)(app_codegen_1.publicPath),
        state,
    };
    return ctx;
}
function includeActions(ctx, actions) {
    const actionNames = Object.keys(actions);
    for (const actionName of actionNames) {
        const action = actions[actionName];
        ctx[actionName] = action.bind(ctx, ctx);
    }
}
exports.includeActions = includeActions;
function createActions(state, events) {
    const context = createContext(state, events);
    includeActions(context, actions);
    return context;
}
exports.createActions = createActions;
