// piral-core/lib/components/PiralRoutes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PiralRoutes = void 0;
const tslib_1 = require("tslib");
const React = tslib_1.__importStar(require("react"));
const hooks_1 = require("../hooks");
const utils_1 = require("../utils");
const app_codegen_1 = require("../../app.codegen");
function useShellRoutes() {
    const routes = (0, hooks_1.useGlobalState)((s) => s.routes);
    return React.useMemo(() => Object.entries(routes).map(([path, Component]) => ({
        path,
        Component,
        meta: Component?.meta || {},
        matcher: (0, utils_1.createRouteMatcher)(path),
    })), [routes]);
}
function usePiletRoutes() {
    const pages = (0, hooks_1.useGlobalState)((s) => s.registry.pages);
    return React.useMemo(() => Object.entries(pages).map(([path, entry]) => ({
        path,
        Component: entry.component,
        meta: entry.meta,
        matcher: (0, utils_1.createRouteMatcher)(path),
    })), [pages]);
}
function useRoutes() {
    const shellRoutes = useShellRoutes();
    const piletRoutes = usePiletRoutes();
    return (0, app_codegen_1.useRouteFilter)([...shellRoutes, ...piletRoutes]);
}
/**
 * The component for defining the exclusive routes to be used.
 */
const PiralRoutes = ({ NotFound, RouteSwitch, ...props }) => {
    const paths = useRoutes();
    return React.createElement(RouteSwitch, { NotFound: NotFound, paths: paths, ...props });
};
exports.PiralRoutes = PiralRoutes;
exports.PiralRoutes.displayName = 'Routes';
